python/lsst/daf/butler/_dataset_type.py

```diff
@@ -323,6 +323,11 @@ class for this dataset type that can convert the python type associated
             or the storage class associated with the other can be converted to
             this.
         """
+        # Support conversion when the components have the same name and
+        # convertible parents, e.g. Parquet storage classes.
+        if self.component() is not None and self.component() == other.component():
+            return self.makeCompositeDatasetType().is_compatible_with(other.makeCompositeDatasetType())
```

Comment:
I don't think this is the right fix. In my testing if I have two composite dataset types that have components that are incompatible this says that everything is fine. We have to test somewhere that the components are compatible.

The parquet test is failing for a different reason. It's failing because `_equal_ignoring_storage_class` is not ignoring parent storage class and so this returns False before the actual storage class is checked.  I think the fix here is to think harder about the parent storage class in this context. Maybe is_compatible_with has to check parent compatibility and component compatibility or maybe it doesn't care about the parent at all.

Reply:
I was thinking that we could declare that two component dataset types are compatible if either of the following is true:
 - if the component storage classes are directly compatible (this is the only supported case on `main`);
 - if the parent storage classes are compatible and the component names are the same (the condition added on this branch, but I think it's implied to already work by the other way the same `get` is already written on the test).

The specific situation here is that as long as `is_compatible_with` doesn't return `False` for `ArrowAstropy.columns` vs. `DataFrame.columns`, the way we call the `Formatter` already works, and we'd rather let the `Formatter` do the conversion anyway.  So I'd rather not have to define explicit conversions between the two component storage classes in this case, since that wouldn't really reflect how the conversion works.

I am somewhat worried that this would lead us to declare two storage classes as convertible when a conversion doesn't actually exist (except in the formatter), but I don't see that coming up in practice in this case so I wasn't sure how much to worry about it.

Reply:
Maybe it would be better if the new check here moved up to the place where we call `is_compatible_with` in `Butler.get`, so that my proposed new rule on component compatibility is not a general one, but rather a chance to let the formatter handle a request where the `Butler` cannot rule out a conversion working.

Reply:
The relevant `columns` storage class definitions already register converters and are compatible (that's the reason why the test above the new test was working since the storage class override on the `butler.get()` was specifying the column storage class).

I think if we start saying that conversions for components don't have to work if the parents can convert that might break the `storageClass=Override` option for components. I don't like the idea of not checking component compatibility at all.

Reply:
I think `is_compatible_with` not checking parent storage class but requiring that components are convertible is good enough isn't it? The test passes now if I remove the parent storage class check. We can assume that if you are comparing components you don't care about the parent.